---
title: "Benchmarking predictors"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

```{r, echo = FALSE}
library(tidyverse)
library(ampir)
library(AmpGram)
library(AmpGramModel)
```

The reference proteomes for the organisms *Homo sapiens* (human), proteome ID: up000005640, and *Arabidopsis thaliana* (mouse-ear cress), proteome ID: up000006548, were downloaded from UniProt proteomes (accessed 23 Jan 2021). *H. sapiens* contained 20,379 reviewed proteins and 55,398 unreviewed proteins. *A. thaliana* contained 15,956 reviewed and 23,390 unreviewed proteins. 


```{r}
human_proteome <- read_tsv("data/proteomes/uniprot-proteome UP000005640.tab") %>% mutate(Label = case_when(str_detect(`Keyword ID`, "KW-0929") ~ "Pos", TRUE ~ "Neg"))
cress_proteome <- read_tsv("data/proteomes/uniprot-proteome UP000006548.tab") %>% mutate(Label = case_when(str_detect(`Keyword ID`, "KW-0929") ~ "Pos", TRUE ~ "Neg"))

reference_proteomes <- rbind(human_proteome, cress_proteome)
```

## ampir 

```{r, eval = FALSE}
cress_pred_ampir_prec <- cress_proteome %>% select(`Entry name`, Sequence) %>% as.data.frame() %>% predict_amps(n_cores = 3) %>% add_column(Organism = "Arabidopsis thaliana") %>% add_column(Model = "ampir_precursor")

cress_pred_ampir_mat <- cress_proteome %>% select(`Entry name`, Sequence) %>% as.data.frame() %>% predict_amps(n_cores = 3, model = "mature") %>% add_column(Organism = "Arabidopsis thaliana") %>% add_column(Model = "ampir_mature")

human_pred_ampir_prec <- human_proteome %>% select(`Entry name`, Sequence) %>% as.data.frame() %>% predict_amps(n_cores = 3) %>% add_column(Organism = "Homo sapiens") %>% add_column(Model = "ampir_precursor")

human_pred_ampir_mat <- human_proteome %>% select(`Entry name`, Sequence) %>% as.data.frame() %>% predict_amps(n_cores = 3, model = "mature") %>% add_column(Organism = "Homo sapiens") %>% add_column(Model = "ampir_mature")
```

In addition to ampir's standard models (`ampir_precursor` and `ampir_mature`), a special model was trained with sequences present in the `ampir_precursor` model, **minus** any *H. sapiens* or *A. thaliana* proteins (see ampir's analysis repository, [AMP_pub](https://github.com/Legana/AMP_pub), workflow 02_build_training_data for details). This was done to remove any potential biases when testing the trained models on the *H. sapiens* and *A. thaliana* proteomes. 

```{r, eval = FALSE}
ampir_prec_model_nobench <- readRDS("data/ampir_v1/tuned_precursor_imbal_nobench.rds")

cress_pred_ampir_nobench <- cress_proteome %>% select(`Entry name`, Sequence) %>% as.data.frame() %>% predict_amps( n_cores=1, model = ampir_prec_model_nobench) %>% add_column(Organism = "Arabidopsis thaliana") %>% add_column(Model = "ampir_precursor_nobench")

human_pred_ampir_nobench <- human_proteome %>% select(`Entry name`, Sequence) %>% as.data.frame() %>% predict_amps( n_cores=1, model = ampir_prec_model_nobench) %>% add_column(Organism = "Homo sapiens") %>% add_column(Model = "ampir_precursor_nobench")
```

```{r}
ampir_proteome_predictions <- rbind(cress_pred_ampir_prec, cress_pred_ampir_mat, cress_pred_ampir_nobench, human_pred_ampir_prec, human_pred_ampir_mat, human_pred_ampir_nobench)
```

```{r, eval = FALSE, echo = FALSE}
saveRDS(cress_pred_ampir_prec, "data/prediction_results/ampir/cress_pred_ampir_prec.rds")
saveRDS(cress_pred_ampir_mat, "data/prediction_results/ampir/cress_pred_ampir_mat.rds")
saveRDS(human_pred_ampir_prec, "data/prediction_results/ampir/human_pred_ampir_prec.rds")
saveRDS(human_pred_ampir_mat, "data/prediction_results/ampir/human_pred_ampir_mat.rds")
saveRDS(cress_pred_ampir_nobench, "data/prediction_results/ampir/cress_pred_ampir_nobench.rds")
saveRDS(ampir_proteome_predictions, "data/prediction_results/ampir/ampir_refproteomes_predictions.rds")
```

```{r, echo = FALSE}
cress_pred_ampir_prec <- readRDS("data/prediction_results/ampir/cress_pred_ampir_prec.rds")
cress_pred_ampir_mat <- readRDS("data/prediction_results/ampir/cress_pred_ampir_mat.rds")
human_pred_ampir_prec <- readRDS("data/prediction_results/ampir/human_pred_ampir_prec.rds")
human_pred_ampir_mat <- readRDS("data/prediction_results/ampir/human_pred_ampir_mat.rds")
cress_pred_ampir_nobench <- readRDS("data/prediction_results/ampir/cress_pred_ampir_nobench.rds")
ampir_proteome_predictions <- readRDS("data/prediction_results/ampir/ampir_refproteomes_predictions.rds")
```

To use the human and *Arabidopsis* proteomes in other predictors, first the non-standard amino acids are removed. This is because the majority of AMP predictors only accept sequences that contain standard amino acids. ampir contains a function that removes everything which is not a standard amino acid, but other predictors, such as AMP scanner, require sequences to be preprocessed to have these sequences removed prior to analysing the sequences. Therefore, the `remove_nonstandard_aa.R` function from ampir is used to remove sequences that contain nonstandard amino acids, and ampir's `df_to_faa.R` function is used to write the files as FASTA files as input to other predictors. In addition, due to sequence amount restrictions using predictor web interfaces, the *Arabidopsis* proteome is split in half and the human proteome is split threefold.  

```{r, eval = FALSE}
arab_prot_clean <- read_faa("data/proteomes/arabidopsis-proteomeUP000006548.fasta") %>% remove_nonstandard_aa() 
df_to_faa(arab_prot_clean, "cache/arab_proteome_standardaa.fasta")

arab_prot_clean %>% slice((1:19669)) %>% df_to_faa("cache/arab_prot_clean1.fasta")
arab_prot_clean %>% slice((19670:n())) %>% df_to_faa("cache/arab_prot_clean2.fasta")

homo_prot_clean <- read_faa("data/proteomes/human-proteomeUP000005640.fasta") %>% remove_nonstandard_aa()
df_to_faa(homo_prot_clean, "cache/homo_proteome_standardaa.fasta")

homo_prot_clean %>% slice((1:22494)) %>% df_to_faa("cache/homo_prot_clean1.fasta")
homo_prot_clean %>% slice((22495:44988)) %>% df_to_faa("cache/homo_prot_clean2.fasta")
homo_prot_clean %>% slice((44989:n())) %>% df_to_faa("cache/homo_prot_clean3.fasta")
```

## Antimicrobial Peptide Scanner vr. 2

AMP scanner vr. 2 (model Feb 2020)


```{r}
ampscanner_file_paths <- c(list.files("data/prediction_results/ampscanner_v2", pattern="*.csv",full.names = T))

ampscan_genome_bench <- do.call(rbind,lapply(ampscanner_file_paths,read_csv)) %>% 
  separate(SeqID,into = c("database","Entry","Entry name"),sep = "\\|") %>% 
  left_join(reference_proteomes,by="Entry") %>% 
  select(ID=Entry,prob_AMP=Prediction_Probability,Organism,Label) %>% 
  add_column(method="ampscannerv2")
```


## AmpGram

```{r}
#do little data first, not working 
#arab_ampgram <- read_txt("cache/arab_proteome_standardaa.fasta")
  
#arab_ampgram_pred <- predict(AmpGram_model, arab_ampgram)


```

