---
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This repository looks at how best to predict AMPs in genomes and how to assess prediction performance

- [Comparison of training and test data used by AMP predictors](01_data_comparison.md) : Rmd file [01_data_comparison.Rmd](01_data_comparison.Rmd)
- [Performance metrics in an omics-scanning context](02_performance_metrics.md) : Rmd file [02_performance_metrics.Rmd](02_performance_metrics.Rmd)
- [Benchmarking AMP predictors](03_benchmarking.md) : Rmd file [03_benchmarking.Rmd](03_benchmarking.Rmd)
- [Taxonomic composition of SwissProt AMPs](04_taxonomic_composition.md) : Rmd file [04_taxonomic_composition.Rmd](04_taxonomic_composition.Rmd)
- [Phylogenetic tree of phyla represented with SwissProt AMPs](04.5_tree.md) : Rmd file [04.5_tree.Rmd](04.5_tree.Rmd)
- [Taxonomic bias effect on model performance](05_taxonomic_bias_effect.md) : Rmd file [05_taxonomic_bias_effect.Rmd](05_taxonomic_bias_effect.Rmd)


The files required to run the code in these Rmd files can be obtained [here](https://cloudstor.aarnet.edu.au/plus/s/duG3rETLzipwW7Z) or by using the command:


```bash
wget 'https://cloudstor.aarnet.edu.au/plus/s/duG3rETLzipwW7Z/download' -O data.tgz
tar -zxvf data.tgz 
```



